MaxText/layers/attentions.py

```diff
@@ -255,9 +257,67 @@ def apply_attention(
                            Use `dot_product` instead."""
         )
       return self.cudnn_flash_attention(query, key, value, decoder_segment_ids, model_mode), None, None
+    elif self.attention_kernel == "pallas_gpu":
+      if model_mode == common_types.MODEL_MODE_AUTOREGRESSIVE:
+        return self.gqa_wrapped_maxtext_compatible(query, key, value)
+      key = jnp.repeat(key, self.num_query_heads // self.num_kv_heads, axis=2)
+      value = jnp.repeat(value, self.num_query_heads // self.num_kv_heads, axis=2)
+      return self.pallas_attention_kernel(query, key, value, decoder_segment_ids, model_mode)
     else:
       raise ValueError(f"Unexpected attention kernel {self.attention_kernel=}.")
 
+  def gqa_wrapped_maxtext_compatible(
+      self,
+      q, k, v,
+      start_idx=None, kv_seq_len=None,
+      sm_scale: float | None = None,
+      block_h: int = 16,
+      block_k: int = 128,
+      k_splits: int = 16,
+      num_warps: int | None = None,
+      num_stages: int = 2,
+      grid: tuple[int, ...] | None = None,
+      interpret: bool = False,
+      debug: bool = False,
+  ):
+    sm_scale = sm_scale if sm_scale is not None else (1 / math.sqrt(q.shape[-1]))
+    batch_size, q_length, q_heads, head_dim = q.shape
+    k_seq_len, kv_heads = k.shape[1], k.shape[2]
+    assert q_heads % kv_heads == 0
+
+    q_reshaped = q.reshape(batch_size, q_length, kv_heads, q_heads // kv_heads, head_dim)
+    k_transposed = jnp.swapaxes(k, 1, 2)
+
+    # Re-compute attention weights to get local_max and local_sum
+    attn_weights = (q_reshaped @ k_transposed.transpose(0, 1, 3, 2)) * sm_scale
+    local_max = jnp.max(attn_weights, axis=-1, keepdims=True)
+    local_exps = jnp.exp(attn_weights - local_max)
+    local_sum = jnp.sum(local_exps, axis=-1, keepdims=True)
```

Comment:
Calculating `local_sum` and `local_max` like this defeats the whole purpose of using a kernel. A better idea is to have the kernel output these values. I'll add it to `jax.experimental.pallas.ops.gpu.pallas_decode_attention`, and then you can remove these lines.

Or better yet, do we absolutely need them? `pallas_attention_kernel` below returns `out, None, None`

Reply:
Gleb, thanks for your input!

Could you please incorporate local_max and local_sum functions into jax.experimental.pallas.ops.gpu.pallas_decode_attention, similar to how it's implemented in this code : 

https://github.com/AI-Hypercomputer/maxtext/blob/main/MaxText/kernels/ragged_attention.py#L348